hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationHandler.java

```diff
@@ -0,0 +1,244 @@
+    try {
+     // State is UNDER_REPLICATED
+      final List<DatanodeDetails> deletionInFlight = new ArrayList<>();
+      for (ContainerReplicaOp op : pendingOpsForContainer) {
+        if (op.getOpType() == ContainerReplicaOp.PendingOpType.DELETE) {
+          deletionInFlight.add(op.getTarget());
+        }
+      }
+      List<Integer> missingIndexes = replicaCount.unavailableIndexes(true);
+      // We got the missing indexes, this is excluded any decommissioning
+      // indexes. Find the good source nodes.
+
+      if (missingIndexes.size() > 0 && !replicaCount.unRecoverable()) {
+        List<ContainerReplica> sources = replicas.stream().filter(r -> r
+            .getState() == StorageContainerDatanodeProtocolProtos
+            .ContainerReplicaProto.State.CLOSED)
+            // Exclude stale and dead nodes. This is particularly important for
+            // maintenance nodes, as the replicas will remain present in the
+            // container manager, even when they go dead.
+            .filter(r -> ReplicationManager
+                .getNodeStatus(r.getDatanodeDetails(), nodeManager).isHealthy())
+            .filter(r -> !deletionInFlight.contains(r.getDatanodeDetails()))
+            .collect(Collectors.toList());
+        LOG.debug("Missing indexes detected for the container {}." +
+                " The missing indexes are {}", id, missingIndexes);
+        // We have source nodes.
+        if (sources.size() >= repConfig.getData()) {
+          final List<DatanodeDetails> selectedDatanodes =
+              getTargetDatanodes(replicas, container, missingIndexes.size());
+
+          List<ReconstructECContainersCommand.DatanodeDetailsAndReplicaIndex>
+              sourceDatanodesWithIndex = new ArrayList<>();
+          for (ContainerReplica srcReplica : sources) {
+            sourceDatanodesWithIndex.add(
+                new ReconstructECContainersCommand
+                    .DatanodeDetailsAndReplicaIndex(
+                    srcReplica.getDatanodeDetails(),
+                    srcReplica.getReplicaIndex()));
+          }
+
+          final ReconstructECContainersCommand reconstructionCommand =
+              new ReconstructECContainersCommand(id.getProtobuf().getId(),
+                  sourceDatanodesWithIndex, selectedDatanodes,
+                  int2byte(missingIndexes),
+                  repConfig);
+          // Keeping the first target node as coordinator.
+          return ImmutableMap.of(selectedDatanodes.get(0),
+              reconstructionCommand);
+        } else {
+          LOG.warn("Cannot proceed for EC container reconstruction for {}, due"
+              + "to insufficient source replicas found. Number of source "
+              + "replicas needed: {}. Number of available source replicas are:"
+              + " {}. Available sources are: {}", container.containerID(),
+              repConfig.getData(), sources.size(), sources);
+        }
+      } else if (result instanceof ContainerHealthResult
```

Comment:
I think we can handle the reconstruction command and the replication commands in a single pass - we don't need it to be `if () ... else if ()` - we should be able to create both sets of commands to repair the container in a single pass.

Reply:
What do you mean by single pass? For replication we iterate and for reconstruction all sources we need to pass to create command. Could you eloborate?

Reply:
We should handle both reconstruction and decommission copies in one call to this method - they are not exclusive so it is not `if...else if`, but `if missing_indexes ... end` and `if decommission_only_indexes()... end`.